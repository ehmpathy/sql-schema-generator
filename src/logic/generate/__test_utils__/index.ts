export * from './databaseConnection';
export * from './createTablesForEntity';
export * from './dropTablesForEntity';
