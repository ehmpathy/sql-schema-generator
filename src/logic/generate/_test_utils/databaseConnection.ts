import mysql, { Connection, ConnectionOptions } from 'mysql2/promise';

export const getDatabaseConnection = async (): Promise<Connection> => {
  const dbConfig: ConnectionOptions = {
    // hardocded since this is just the integration test db config; settings configured in the provisioning/.../docker.file
    host: 'localhost',
    user: 'root',
    password: 'a-secure-password', // its just a docker integration test db, no worries about hardcoding this
    database: 'superimportantdb',
    port: 12821,
  };
  const connection = await mysql.createConnection(dbConfig);
  return connection;
};

export { Connection as DatabaseConnection };
