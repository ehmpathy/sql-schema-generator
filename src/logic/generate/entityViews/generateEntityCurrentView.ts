import indentString from 'indent-string';

import { Entity } from '../../../domain';
import { castPropertyToSelector } from './utils/castPropertyToSelector';

/*
  SELECT
    s.id,
    s.uuid,
    s.[all static props] (concat_ws all array props)
    v.[all updateable props] (concat_ws all array props)
    s.created_at
    v.effective_at
    v.created_at as updated_at
*/

export const generateEntityCurrentView = ({ entity }: { entity: Entity }) => {
  // define basics
  const viewName = `view_${entity.name}_current`;

  // define the property names
  const staticPropertyNames = Object.entries(entity.properties)
    .filter((entry) => !entry[1].updatable)
    .map((entry) => entry[0]);
  const updateablePropertyNames = Object.entries(entity.properties)
    .filter((entry) => !!entry[1].updatable)
    .map((entry) => entry[0]);
  const arrayPropertyNames = Object.entries(entity.properties)
    .filter((entry) => !!entry[1].array)
    .map((entry) => entry[0]);

  // if there are no updateable properties and no array properties, don't return a view - it would just be a replica of the static table
  const hasVersionTable = !!updateablePropertyNames.length;
  const hasArrayProperties = !!arrayPropertyNames.length;
  if (!hasVersionTable && !hasArrayProperties) return null;

  // define selectors
  const staticPropertySelectors = staticPropertyNames.map((name) =>
    castPropertyToSelector({
      entityName: entity.name,
      name,
      definition: entity.properties[name],
    }),
  );
  const updateablePropertySelectors = updateablePropertyNames.map((name) =>
    castPropertyToSelector({
      entityName: entity.name,
      name,
      definition: entity.properties[name],
    }),
  );

  // define the columns
  const columns = [
    's.id',
    's.uuid',
    ...staticPropertySelectors,
    ...updateablePropertySelectors,
    's.created_at',
    hasVersionTable ? 'v.effective_at' : null,
    hasVersionTable ? 'v.created_at as updated_at' : null,
  ].filter((column) => !!column); // only non empty columns

  // define the joins
  const joins = hasVersionTable // only join if has version table
    ? [
        '',
        `JOIN ${entity.name}_cvp cvp ON s.id = cvp.${entity.name}_id`,
        `JOIN ${entity.name}_version v ON v.id = cvp.${entity.name}_version_id`,
      ]
    : [];

  // combine the version and static logic into full upsert function
  const definition = `
CREATE OR REPLACE VIEW ${viewName} AS
  SELECT
    ${indentString(columns.join(',\n'), 4).trim()}
  FROM ${entity.name} s${joins.join('\n  ')};
  `.trim();
  return {
    name: viewName,
    sql: definition,
  };
};
