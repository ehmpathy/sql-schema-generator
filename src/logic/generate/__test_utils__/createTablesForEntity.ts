import { Entity } from '../../../types';
import { generateEntityTables } from '../entityTables/generateEntityTables';
import { DatabaseConnection } from './databaseConnection';

export const createTablesForEntity = async ({
  entity,
  dbConnection,
}: {
  entity: Entity;
  dbConnection: DatabaseConnection;
}) => {
  const tables = await generateEntityTables({ entity });

  // create the static table
  await dbConnection.query({ sql: tables.static.sql });

  // drop the version table, if exists
  if (tables.version) {
    await dbConnection.query({ sql: tables.version.sql });
  }

  // drop the cvp table, if exists
  if (tables.currentVersionPointer) {
    await dbConnection.query({ sql: tables.currentVersionPointer.sql });
  }

  // create each mapping table
  await Promise.all(tables.mappings.map((mappingTable) => dbConnection.query({ sql: mappingTable.sql })));
};
