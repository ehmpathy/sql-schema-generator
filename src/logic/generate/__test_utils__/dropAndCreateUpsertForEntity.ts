import { Entity } from '../../../types';
import { generateEntityUpsert } from '../entityFunctions/generateEntityUpsert';
import { DatabaseConnection } from './databaseConnection';
import { provisionGetFromDelimiterSplitStringFunction } from './provisionGetFromDelimiterSplitStringFunction';

export const dropAndCreateUpsertFunctionForEntity = async ({
  entity,
  dbConnection,
}: {
  entity: Entity;
  dbConnection: DatabaseConnection;
}) => {
  // if has array properties, provision the array access function
  if (Object.values(entity.properties).some((prop) => !!prop.array)) {
    await provisionGetFromDelimiterSplitStringFunction({ dbConnection });
  }

  // drop and create the upsert
  const { name, sql: upsertSql } = generateEntityUpsert({ entity });
  await dbConnection.query({ sql: `DROP FUNCTION IF EXISTS ${name}` });
  await dbConnection.query({ sql: upsertSql });
};
