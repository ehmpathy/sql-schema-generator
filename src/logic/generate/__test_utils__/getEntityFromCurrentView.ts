import { Entity } from '../../../types';
import { generateEntityCurrentView } from '../entityViews/generateEntityCurrentView';
import { DatabaseConnection } from './databaseConnection';

export const getEntityFromCurrentView = async ({
  id,
  entity,
  dbConnection,
}: {
  id: number;
  entity: Entity;
  dbConnection: DatabaseConnection;
}) => {
  const view = generateEntityCurrentView({ entity });
  const name = view!.name;
  const results = (await dbConnection.execute({ sql: `select * from ${name} where id = ${id}` })) as any;
  expect(results[0].length).toEqual(1);
  const entityData = results[0][0];
  return entityData;
};
