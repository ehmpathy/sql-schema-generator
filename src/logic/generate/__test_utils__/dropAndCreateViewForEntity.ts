import { DatabaseConnection } from '../../../__test_utils__/databaseConnection';
import { Entity } from '../../../types';
import { generateEntityCurrentView } from '../entityViews/generateEntityCurrentView';

export const dropAndCreateViewForEntity = async ({
  entity,
  dbConnection,
}: {
  entity: Entity;
  dbConnection: DatabaseConnection;
}) => {
  const { name, sql: upsertSql } = generateEntityCurrentView({ entity })!;
  await dbConnection.query({ sql: `DROP VIEW IF EXISTS ${name}` });
  await dbConnection.query({ sql: upsertSql });
  return { name, sql: upsertSql };
};
