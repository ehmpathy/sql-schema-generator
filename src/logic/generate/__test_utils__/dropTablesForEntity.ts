import { DatabaseConnection } from '../../../__test_utils__/databaseConnection';
import { Entity } from '../../../domain';
import { generateEntityTables } from '../entityTables/generateEntityTables';

export const dropTablesForEntity = async ({
  entity,
  dbConnection,
}: {
  entity: Entity;
  dbConnection: DatabaseConnection;
}) => {
  const tables = await generateEntityTables({ entity });

  // drop each mapping table
  await Promise.all(
    tables.mappings.map((mappingTable) =>
      dbConnection.query({ sql: `DROP TABLE IF EXISTS ${mappingTable.name} CASCADE;` }),
    ),
  );

  // drop the cvp table, if exists
  if (tables.currentVersionPointer) {
    await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.currentVersionPointer.name} CASCADE;` });
  }

  // drop the version table, if exists
  if (tables.version) {
    await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.version!.name} CASCADE;` });
  }

  // drop the static table
  await dbConnection.query({ sql: `DROP TABLE IF EXISTS ${tables.static.name} CASCADE;` });
};
