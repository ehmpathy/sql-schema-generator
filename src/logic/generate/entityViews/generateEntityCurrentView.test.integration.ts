import uuid from 'uuid/v4';
import { pg as prepare } from 'yesql';

import { DatabaseConnection, getDatabaseConnection } from '../../../__test_utils__/databaseConnection';
import { Entity, ValueObject } from '../../../types';
import * as prop from '../../define/defineProperty';
import { createTablesForEntity, dropTablesForEntity } from '../__test_utils__';
import { dropAndCreateUpsertFunctionForEntity } from '../__test_utils__/dropAndCreateUpsertForEntity';
import { dropAndCreateViewForEntity } from '../__test_utils__/dropAndCreateViewForEntity';
import { getEntityFromCurrentView } from '../__test_utils__/getEntityFromCurrentView';
import { generateEntityCurrentView } from './generateEntityCurrentView';

describe('generateEntityViewCurrent', () => {
  let dbConnection: DatabaseConnection;
  beforeAll(async () => {
    dbConnection = await getDatabaseConnection();
  });
  afterAll(async () => {
    await dbConnection.end();
  });
  describe('static entity', () => {
    describe('without array properties', () => {
      const address = new Entity({
        name: 'alternative_address',
        properties: {
          street: prop.VARCHAR(255),
          suite: {
            ...prop.VARCHAR(255),
            nullable: true,
          },
          city: prop.VARCHAR(255),
          country: prop.ENUM(['US', 'CA', 'MX']),
          weekday_found: {
            // non-unique but static property -> only track the first value
            ...prop.VARCHAR(15),
            nullable: true,
          },
        },
        unique: ['street', 'suite', 'city', 'country'],
      });
      it('should not create a _current view for static entities without array properties', async () => {
        const view = generateEntityCurrentView({ entity: address });
        expect(view).toEqual(null);
      });
    });
    describe('with array properties', () => {
      const lock = new ValueObject({
        name: 'door_lock',
        properties: {
          manufacturer: prop.VARCHAR(255),
          manufacturerId: prop.VARCHAR(255),
        },
      });
      const door = new ValueObject({
        name: 'door',
        properties: {
          color: prop.ENUM(['red', 'green', 'blue']),
          lock_ids: prop.ARRAY_OF(prop.REFERENCES(lock)), // e.g., can have one lock or two locks
        },
      });
      const upsertLock = async ({ manufacturer, manufacturerId }: { manufacturer: string; manufacturerId: string }) => {
        const result = await dbConnection.query(
          prepare(`
            SELECT * FROM upsert_${lock.name}(
              :manufacturer,
              :manufacturerId
            );
        `)({
            manufacturer,
            manufacturerId,
          }),
        );
        return result.rows[0].id;
      };
      const upsertDoor = async ({ color, lock_ids }: { color: string; lock_ids: string }) => {
        const result = await dbConnection.query(
          prepare(`
            SELECT * FROM upsert_${door.name}(
              :color,
              :lock_ids
            );
        `)({
            color,
            lock_ids,
          }),
        );
        return result.rows[0].id;
      };
      const getEntityFromView = async ({ id }: { id: number }) =>
        getEntityFromCurrentView({ id, entity: door, dbConnection });
      beforeAll(async () => {
        await dropTablesForEntity({ entity: door, dbConnection });
        await dropTablesForEntity({ entity: lock, dbConnection });
        await createTablesForEntity({ entity: lock, dbConnection });
        await createTablesForEntity({ entity: door, dbConnection });
        await dropAndCreateUpsertFunctionForEntity({ entity: lock, dbConnection });
        await dropAndCreateUpsertFunctionForEntity({ entity: door, dbConnection });
      });
      it('should generate good looking and consistent sql', () => {
        const view = generateEntityCurrentView({ entity: door });
        expect(view).not.toEqual(null);
        expect(view!.sql).toMatchSnapshot(); // review this manually for changes
      });

      it('should show the entity accurately', async () => {
        await dropAndCreateViewForEntity({ entity: door, dbConnection });
        const lockIds = [
          await upsertLock({ manufacturer: uuid(), manufacturerId: uuid() }),
          await upsertLock({ manufacturer: uuid(), manufacturerId: uuid() }),
          await upsertLock({ manufacturer: uuid(), manufacturerId: uuid() }),
        ].join(',');

        const props = {
          color: 'red',
          lock_ids: `{${lockIds}}`,
        };
        const id = await upsertDoor(props);

        // // check that the static part was accurate
        const entity = await getEntityFromView({ id });
        expect({ ...entity, lock_ids: `{${entity.lock_ids.join(',')}}` }).toMatchObject(props);
        expect(entity.uuid.length).toEqual(36); // sanity check that its a uuid
        expect(entity.id).toEqual(id);
      });
      it('should still return an array even if an array property for the entity is empty', async () => {
        await dropAndCreateViewForEntity({ entity: door, dbConnection });

        const props = {
          color: 'red',
          lock_ids: '{}',
        };
        const id = await upsertDoor(props);

        // // check that the static part was accurate
        const entity = await getEntityFromView({ id });
        expect(Array.isArray(entity.lock_ids));
        expect(entity.lock_ids).toEqual([]);
      });
    });
  });
  describe('versioned entity', () => {
    describe('without array properties', () => {
      const user = new Entity({
        name: 'alternative_user',
        properties: {
          cognito_uuid: prop.UUID(),
          name: {
            ...prop.VARCHAR(255),
            updatable: true,
          },
          bio: {
            ...prop.TEXT(),
            updatable: true,
            nullable: true,
          },
        },
        unique: ['cognito_uuid'],
      });
      beforeAll(async () => {
        await dropTablesForEntity({ entity: user, dbConnection });
        await createTablesForEntity({ entity: user, dbConnection });
        await dropAndCreateUpsertFunctionForEntity({ entity: user, dbConnection });
      });
      const getEntityFromView = async ({ id }: { id: number }) =>
        getEntityFromCurrentView({ id, entity: user, dbConnection });
      const upsertUser = async ({ cognito_uuid, name, bio }: { cognito_uuid: string; name: string; bio?: string }) => {
        const result = await dbConnection.query(
          prepare(`
        SELECT * FROM upsert_${user.name}(
          :cognito_uuid,
          :name,
          :bio
        );
      `)({
            cognito_uuid,
            name,
            bio,
          }),
        );
        return result.rows[0].id;
      };
      it('should generate good looking and consistent sql', () => {
        const view = generateEntityCurrentView({ entity: user });
        expect(view).not.toEqual(null);
        expect(view!.sql).toMatchSnapshot(); // review this manually for changes
      });
      it('should show the entity accurately', async () => {
        await dropAndCreateViewForEntity({ entity: user, dbConnection });
        const props = {
          cognito_uuid: uuid(),
          name: 'hank hill',
          bio: 'i sell propane and propane accessories',
        };
        const id = await upsertUser(props);

        // check that the static part was accurate
        const entity = await getEntityFromView({ id });
        expect(entity).toMatchObject(props);
        expect(entity.uuid.length).toEqual(36); // sanity check that its a uuid
        expect(entity.id).toEqual(id);
      });
      it('should always show the current version', async () => {
        await dropAndCreateViewForEntity({ entity: user, dbConnection });
        const props = {
          cognito_uuid: uuid(),
          name: 'hank hill',
          bio: 'i sell propane and propane accessories',
        };
        const id = await upsertUser(props);
        const idAgain = await upsertUser({ ...props, name: 'Hank Hillerson' });
        await upsertUser({ ...props, name: 'Hank Hillbody', bio: undefined });
        expect(idAgain).toEqual(id);

        // check that the updated part is still accurate
        const entity = await getEntityFromView({ id });
        expect(entity).toMatchObject({ ...props, name: 'Hank Hillbody', bio: null });
      });
    });
    describe('with array properties', () => {
      const wheel = new ValueObject({
        name: 'wheel',
        properties: {
          name: prop.VARCHAR(255),
        },
      });
      const vehicle = new Entity({
        name: 'vehicle',
        properties: {
          name: prop.VARCHAR(255),
          wheel_ids: {
            ...prop.ARRAY_OF(prop.REFERENCES(wheel)),
            updatable: true, // the wheels on a vehicle can change
          },
        },
        unique: ['name'],
      });
      beforeAll(async () => {
        await dropTablesForEntity({ entity: vehicle, dbConnection });
        await dropTablesForEntity({ entity: wheel, dbConnection });
        await createTablesForEntity({ entity: wheel, dbConnection });
        await createTablesForEntity({ entity: vehicle, dbConnection });

        await dropAndCreateUpsertFunctionForEntity({ entity: wheel, dbConnection });
        await dropAndCreateUpsertFunctionForEntity({ entity: vehicle, dbConnection });
      });
      const getEntityFromView = async ({ id }: { id: number }) =>
        getEntityFromCurrentView({ id, entity: vehicle, dbConnection });
      const upsertVehicle = async ({ name, wheel_ids }: { name: string; wheel_ids: string }) => {
        const result = await dbConnection.query(
          prepare(`
          SELECT * FROM upsert_${vehicle.name}(
            :name,
            :wheel_ids
          );
        `)({
            name,
            wheel_ids,
          }),
        );
        return result.rows[0].id;
      };
      const upsertWheel = async ({ name }: { name: string }) => {
        const result = await dbConnection.query(
          prepare(`
          SELECT * FROM upsert_${wheel.name}(
            :name
          );
        `)({
            name,
          }),
        );
        return result.rows[0].id;
      };
      it('should generate good looking and consistent sql', () => {
        const view = generateEntityCurrentView({ entity: vehicle });
        expect(view).not.toEqual(null);
        expect(view!.sql).toMatchSnapshot(); // review this manually for changes
      });
      it('should show the entity accurately', async () => {
        await dropAndCreateViewForEntity({ entity: vehicle, dbConnection });
        const wheelIds = [
          await upsertWheel({ name: uuid() }),
          await upsertWheel({ name: uuid() }),
          await upsertWheel({ name: uuid() }),
        ].join(',');
        const props = {
          name: uuid(),
          wheel_ids: `{${wheelIds}}`,
        };
        const id = await upsertVehicle(props);

        // check that the static part was accurate
        const entity = await getEntityFromView({ id });
        expect({ ...entity, wheel_ids: `{${entity.wheel_ids.join(',')}}` }).toMatchObject(props);
        expect(entity.uuid.length).toEqual(36); // sanity check that its a uuid
        expect(entity.id).toEqual(id);
      });
      it('should always show the current version', async () => {
        await dropAndCreateViewForEntity({ entity: vehicle, dbConnection });

        // create original version of entity
        const wheelIds = [
          await upsertWheel({ name: uuid() }),
          await upsertWheel({ name: uuid() }),
          await upsertWheel({ name: uuid() }),
        ].join(',');
        const props = {
          name: uuid(),
          wheel_ids: `{${wheelIds}}`,
        };
        const id = await upsertVehicle(props);

        // update the entity dynamic properties
        const updatedProps = {
          ...props,
          wheel_ids: `{${wheelIds
            .split(',')
            .slice(1, 2)
            .join(',')}}`,
        };
        const idAgain = await upsertVehicle(updatedProps);
        expect(idAgain).toEqual(id);

        // check that the updated part is still accurate
        const entity = await getEntityFromView({ id });
        expect({ ...entity, wheel_ids: `{${entity.wheel_ids.join(',')}}` }).toMatchObject(updatedProps);
      });
    });
  });
});
