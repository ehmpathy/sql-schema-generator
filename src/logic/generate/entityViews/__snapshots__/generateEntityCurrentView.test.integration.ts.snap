// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateEntityViewCurrent static entity with array properties should generate good looking and consistent sql 1`] = `
"CREATE VIEW \`view_door_current\` AS
  SELECT
    s.id,
    s.uuid,
    s.color,
    (
      SELECT GROUP_CONCAT(door_to_door_lock.door_lock_id ORDER BY door_to_door_lock.door_lock_id)
      FROM door_to_door_lock WHERE door_to_door_lock.door_id = s.id
    ) as lock_ids,
    s.created_at
  FROM door s
  ;"
`;

exports[`generateEntityViewCurrent versioned entity with array properties should generate good looking and consistent sql 1`] = `
"CREATE VIEW \`view_vehicle_current\` AS
  SELECT
    s.id,
    s.uuid,
    s.name,
    (
      SELECT GROUP_CONCAT(vehicle_version_to_wheel.wheel_id ORDER BY vehicle_version_to_wheel.wheel_id)
      FROM vehicle_version_to_wheel WHERE vehicle_version_to_wheel.vehicle_version_id = v.id
    ) as wheel_ids,
    s.created_at,
    v.effective_at,
    v.created_at as updated_at
  FROM vehicle s
  JOIN vehicle_cvp cvp ON s.id = cvp.vehicle_id
  JOIN vehicle_version v ON v.id = cvp.vehicle_version_id;"
`;

exports[`generateEntityViewCurrent versioned entity without array properties should generate good looking and consistent sql 1`] = `
"CREATE VIEW \`view_alternative_user_current\` AS
  SELECT
    s.id,
    s.uuid,
    s.cognito_uuid,
    v.name,
    v.bio,
    s.created_at,
    v.effective_at,
    v.created_at as updated_at
  FROM alternative_user s
  JOIN alternative_user_cvp cvp ON s.id = cvp.alternative_user_id
  JOIN alternative_user_version v ON v.id = cvp.alternative_user_version_id;"
`;
